### ISSUE TITLE
Generated Markdown is missing line breaks between heading and image

### ISSUE THREAD

**Issue by kiejo** [syntax-tree/mdast-util-to-markdown#59]
### Initial checklist

- [X] I read the [support docs](https://github.com/syntax-tree/.github/blob/main/support.md)
- [X] I read the [contributing guide](https://github.com/syntax-tree/.github/blob/main/contributing.md)
- [X] I agree to follow the [code of conduct](https://github.com/syntax-tree/.github/blob/main/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Asyntax-tree&type=Issues) and couldn’t find anything (or linked relevant results below)

### Affected packages and versions

mdast-util-to-markdown 1.5.0

### Steps to reproduce

```javascript
import {toMarkdown} from 'mdast-util-to-markdown'

const tree = {
  type: 'root',
  children: [
    { type: 'heading', depth: 1, children: [{ type: 'text', value: 'Test' }] },
    {
      type: 'image',
      url: 'https://example.com/image.png',
    },
  ]
}

console.log(toMarkdown(tree))
```

### Expected behavior

It should log the following output:

```
# Test

![](https://example.com/image.png)
```

### Actual behavior

It logs the following output:

```
# Test![](https://example.com/image.png)
```

It looks like this issue was introduced with this commit: https://github.com/syntax-tree/mdast-util-to-markdown/commit/122101f19c9c7c1f7cc24b66d23608b00fa177af

### Affected runtime and version

node@16.13.2

**Comment by wooorm** [comment]
Hi! Please follow the `mdast` specification, and wrap your images in paragraphs. They may not appear on their own like that

Duplicate of GH-58

**Comment by wooorm** [comment]
Duplicate of #58

**Comment by kiejo** [comment]
Thanks for the quick response! I can confirm that wrapping the image in a paragraph solves the issue, but I noticed that this causes the image alt text to get escaped in a way I did not expect. I would have expected the escaping of the image alt text to be treated in a similar way to the title text.

While this sounds like a separate issue, I thought it would make sense to bring it up in this context as I did not encounter this issue when not wrapping the image inside a paragraph (which I now understand is not spec compliant).

Here is an example for reference:
```javascript
import {toMarkdown} from 'mdast-util-to-markdown'

const tree = {
  type: 'root',
  children: [
    { type: 'paragraph', children: [
      {
        type: 'image',
        url: 'https://example.com/image.png',
        alt: '*hello* [world]',
        title: '*hello* "world"',
      }
    ]},
  ]
}

console.log(toMarkdown(tree))
```

Output:
```
![\*hello\* \[world\]](https://example.com/image.png "*hello* \"world\"")
```

Expected:
```
![*hello* [world\]](https://example.com/image.png "*hello* \"world\"")
```

Please let me know if the current image alt text escaping is by design or if I should open a separate issue for this.

**Comment by wooorm** [comment]
Your expected markdown doesn’t work. Try pasting here:

```markdown
![\*hello\* \[world\]](https://example.com/image.png "*hello* \"world\"")

![*hello* [world\]](https://example.com/image.png "*hello* \"world\"")
```

![\*hello\* \[world\]](https://example.com/image.png "*hello* \"world\"")

![*hello* [world\]](https://example.com/image.png "*hello* \"world\"")

---

> Please let me know if the current image alt text escaping is by design or if I should open a separate issue for this.

Escaping is definitely by design. That being sad, as this case shows, escaping is incredibly hard. If it can be improved, that’s of course great, but maybe it can’t.

**Comment by wooorm** [comment]
If this is your reduced test case: 

```markdown
![*alpha*](xxx) and ![\*bravo\*](yyy)
```

you should check out the `alt` text of the resulting HTML:

![*alpha*](xxx) and ![\*bravo\*](yyy)

---

Note that the `*`s disappear. That’s because they actually are interpreted in images. Markdown “works” in there. But, the tags (`<em>` and `</em>` in this case) are ignored.

**Comment by kiejo** [comment]
Thanks for the additional details and examples. I did not expect image alt text to be interpreted as Markdown as the [`Alternative`](https://github.com/syntax-tree/mdast#alternative) mixin only specifies a simple `string` attribute just like the `title` attribute is defined as part of the [`Resource`](https://github.com/syntax-tree/mdast#resource) mixin.

While the `alt` and `title` attributes seem to be using the same definition, they are interpreted in different ways:
```javascript
import {fromMarkdown} from 'mdast-util-from-markdown'
console.log(JSON.stringify(fromMarkdown('![*alt*](xxx "*title*")'), null, 2))
```
Output of image node:
```json
{
  "type": "image",
  "title": "*title*",
  "url": "xxx",
  "alt": "alt"
}

```

I am wondering if it actually makes sense to interpret the image alt text as Markdown if the information gets discarded and is not even part of the generated AST. The only use case I could come up with so far, would be for Markdown syntax highlighting where the alt text of an image could be rendered in a stylized way. But when for example trying this in Visual Studio Code, the image alt text does not seem to be interpreted as Markdown and instead as simple text (which also matches my initial expectation):
![img-alt-md](https://user-images.githubusercontent.com/152507/216346211-2539d22d-d211-4322-832c-9d95cfefe935.png)

As you can see in the screenshot, the alt text of the image does not use italics or bold while the content of the link does.

At the same time it looks like the current logic matches GitHub's behavior, which might be reason enough to keep it as is.

**Comment by wooorm** [comment]
This isn’t about `mdast`, which is a specification for representing markdown as nodes. And, yeah, there these fields both use `string` as types.

This is about how markdown works. Markdown prescribes this. CommonMark-compliant markdown processors do this.

On the parsing side, this is because links, images, but also emphasis, are not parsed the way you think.
They are parsed in chunks: `*` is seen. Or `![` is seen. And `**` is seen. And then `](#whatever)`.

If such a valid closing is found (e.g., `]`). *Then* the parser looks back to see if there is a corresponding opening (e.g., `![`).
Meanwhile, emphasis/strong (what I call attention) is parsed as separate runs around all that (`*`, `**`, `***`).
And then “normal” things are parsed as whole units from left to right, such as `&amp;`, or `<https://example.com>`, or `` `asd` ``.

So you see, this parsing is already done. Only afterwards do we know that these things occurred in an image.
And then CommonMark says: well, `<em>` in there doesn’t make sense, drop those tags.

**Comment by wooorm** [comment]
More on how parsing works in https://github.com/wooorm/markdown-rs/blob/main/src/construct/label_start_image.rs and https://github.com/wooorm/markdown-rs/blob/main/src/construct/label_end.rs.
A related issue on CommonMark is this one: https://github.com/commonmark/commonmark-spec/issues/716.

**Comment by wooorm** [comment]
:)

A bit about the spec: it’s also a bit of a vague document. It’s nice that it provides many examples, more than we had before, but there are still many edge cases. So we still need to look at what parsers actually do: how does GFM for example *actually* work?

---

Yeah, I’ve thought about adding `children` to images. Which would include stuff like code in there and such. The downside is that you then have two sources of truths. A plugin would edit one, but would likely forget about the other, or at least it’ll get out of sync. And that’s what an AST is for, to remove such complexity, to make it simple to change things.

One alternative is to remove `alt` completely, and only serialize this stuff when going to html. But that’s rather breaking.

---

Perhaps tho, you have interesting needs, for such rich content. By hearing them I might be able to advise better? E.g., perhaps you can use [directives](https://github.com/remarkjs/remark-directive), which do allow rich content, and compile to some other element that has a `figcaption` or so?

**Comment by kiejo** [comment]
I agree with the downside of having two sources of truths and I like the simplicity of the current AST.

My needs are actually quite simple, which is why I would personally prefer the image description to simply work like the `alt` attribute of an HTML `img` tag without characters like "*" or "_" getting dropped/interpreted or needing to escape them. It was quite unexpected to me that the image description supports all these inline elements in the spec as I have always thought of it as the `alt` attribute of an HTML image. With this mindset, using directives for more advanced needs makes a lot of sense, just like you suggested.

I guess the goal of the current image description spec was to make it consistent with the way links work as the syntax is so similar. So I think I understand the trade-off that was being made in the spec even if there might not be that many good use cases for all these inline elements in the image description. Or maybe there are common use cases that I just haven't encountered yet.

After having thought through several use cases and now having a better understanding of the CommonMark spec, I think you picked really good trade-offs for this library. The result is a spec compliant system, which hides some of the flexibility/complexity of CommonMark image descriptions while making it very easy to use for the 99% of cases where a simple `alt` attribute is all the user needs and expects. Great work!

Thanks so much for taking the time to answer all my questions and helping me better understand the design decisions of this library :)
